File: complex_net.py
import torch
import torch.nn as nn
from torch.autograd import Variable


class Branch(nn.Module):
    def __init__(self,b2):
        """Constructs each branch necessary depending on input
    
        Args: 
            b2(nn.Module()): An nn.Conv2d() is passed with specific params
      
        """
        super(Branch, self).__init__()
        self.b = nn.MaxPool2d(kernel_size=2, stride=2)
        self.b2 = b2
        
    def forward(self,x):
        x = self.b(x) 
        y = [self.b2(x).view(-1), self.b2(x).view(-1)]
        z = torch.cat((y[0],y[1]))
        return z
        
        
class ComplexNet(nn.Module):
    def __init__(self, m1, m2):
        """Constructs the base of the network and attaches the branches

        Args: 
            m1(nn.Sequential()): The first segment of the base network
            m2(nn.Sequential()): The second segment of the base network

        """
        super(ComplexNet, self).__init__()
        self.net1 = m1
        self.net2 = m2
        self.net3 = nn.Conv2d(128,256,kernel_size=3,padding=1)
        self.branch1 = Branch(nn.Conv2d(64,64,kernel_size=3,padding=1))
        self.branch2 = Branch(nn.Conv2d(128,256,kernel_size=3, padding=1))
         
    def forward(self, x):
        x = self.net1(x)
        x1 = self.branch1(x)
        y = self.net2(x)
        x2 = self.branch2(y)
        x3 = self.net3(y).view(-1)
        output = torch.cat((x1,x2,x3),0)
        return output
    
def make_layers(params, ch): 
    """Constructs the base segments of the network and returns as nn.Sequential()

    Args: 
        params(int[]): The Conv2d parameters for input and output channels
        ch(int): The initial input channel parameter for the first Conv2d

    """

    layers = []
    channels = ch
    for p in params:
            conv2d = nn.Conv2d(channels, p, kernel_size=3, padding=1)
            layers += [conv2d, nn.ReLU(inplace=True)]
            channels = p
    return nn.Sequential(*layers) 
   
net = ComplexNet(make_layers([64,64],3),make_layers([128,128],64))
return net

File: complex_net.lua
require 'nn'

--[[
This file contains the code definition of our complex example network in raw Torch
using nn.Table modules.  Since nngraph is not used, this network is defined in
reverse order in terms of sequential segments and branches.
]]--


-- the final layer of our base network 
net3 = nn.Sequential()
net3:add(nn.SpatialConvolution(128,256,3,3,1,1,1,1))
net3:add(nn.View(-1))

-- split the last branch 
branch2b = nn.ParallelTable()
branch2b:add(nn.SpatialConvolution(128,256,3,3,1,1,1,1))
branch2b:add(nn.SpatialConvolution(128,256,3,3,1,1,1,1))

-- here we add max pooling and use nn.Replicate() --> nn.SplitTable
-- to allow for a ParallelTable to split the branch 
branch2a = nn.Sequential()
branch2a:add(nn.SpatialMaxPooling(2,2,2,2))
branch2a:add(nn.Replicate(4))
branch2a:add(nn.SplitTable(1))
branch2a:add(branch2b)

-- second (final) branch from base of network
branch2 = nn.ParallelTable()
branch2:add(branch2a)
branch2:add(net3)

-- using nn.Sequential() for part 2 of base network 
net2 = nn.Sequential() 
net2:add(nn.SpatialConvolution(64,128,3,3,1,1,1,1))
net2:add(nn.ReLU())
net2:add(nn.SpatialConvolution(128,128,3,3,1,1,1,1))
net2:add(nn.ReLU())
net2:add(nn.Replicate(4))
net2:add(nn.SplitTable(1))
net2:add(branch2)

-- split from first branch 
branch1b = nn.ParallelTable()
branch1b:add(nn.SpatialConvolution(64,64,3,3,1,1,1,1))
branch1b:add(nn.SpatialConvolution(64,64,3,3,1,1,1,1))

-- add a max pooling layer to branch and then prepare to split 
branch1a = nn.Sequential()
branch1a:add(nn.SpatialMaxPooling(2,2,2,2))
branch1a:add(nn.Replicate(4))
branch1a:add(nn.SplitTable(1))
branch1a:add(branch1b)

-- first branch from the base of the network
branch1 = nn.ParallelTable()
branch1:add(branch1a)
branch1:add(net2)

-- the first layers of the network 
net = nn.Sequential()
net:add(nn.SpatialConvolution(3,64,3,3,1,1,1,1))
net:add(nn.ReLU())
net:add(nn.SpatialConvolution(64,64,3,3,1,1,1,1))
net:add(nn.ReLU())
net:add(nn.Replicate(4))
net:add(nn.SplitTable(1))
net:add(branch1)
